tools/smithy-rs-sync/src/main.rs

```diff
@@ -0,0 +1,409 @@
+            branch,
+        )
+    })?;
+
+    let number_of_commits = commit_revs.len();
+    println!(
+        "Found {} unsynced commit(s), syncing now...",
+        number_of_commits
+    );
+    // Run through all the new commits, syncing them one by one
+    for (i, rev) in commit_revs.iter().enumerate() {
+        let commit = smithy_rs_repo
+            .find_commit(*rev)
+            .with_context(|| format!("couldn't find commit {} in smithy-rs", rev))?;
+
+        println!("[{}/{}]\tsyncing {}...", i + 1, number_of_commits, rev);
+        checkout_commit_to_sync_from(&smithy_rs_repo, &commit).with_context(|| {
+            format!(
+                "couldn't checkout commit {} from smithy-rs that we needed for syncing",
+                rev,
+            )
+        })?;
+
+        let build_artifacts = build_sdk(smithy_rs).context("couldn't build SDK")?;
+        clean_out_existing_sdk(aws_sdk)
+            .context("couldn't clean out existing SDK from aws-sdk-rust")?;
+        copy_sdk(&build_artifacts, aws_sdk)?;
+        create_mirror_commit(&aws_sdk_repo, &commit)
+            .context("couldn't commit SDK changes to aws-sdk-rust")?;
+    }
+    println!("Successfully synced {} commit(s)", commit_revs.len());
+
+    // Get the last commit we synced so that we can set that for the next time this tool gets run
+    let last_synced_commit = commit_revs
+        .last()
+        .expect("can't be empty because we'd have early returned");
+    println!("Updating 'commit at last sync' to {}", last_synced_commit);
+
+    // Update the file containing the commit hash
+    set_last_synced_commit(&aws_sdk_repo, last_synced_commit).with_context(|| {
+        format!(
+            "couldn't write last synced commit hash ({}) to aws-sdk-rust/{}",
+            last_synced_commit, COMMIT_HASH_FILENAME,
+        )
+    })?;
+    // Commit the file containg the commit hash
+    commit_last_synced_commit_file(&aws_sdk_repo)
+        .context("couldn't commit the last synced commit hash file to aws-sdk-rust")?;
+
+    println!(
+        "Successfully synced all mirror commits to aws-sdk-rust/{}. Don't forget to push them",
+        branch
+    );
+
+    Ok(())
+}
+
+/// Starting from a given commit, walk the tree to its `HEAD` in order to build a list of commits that we'll
+/// need to sync. If you don't see the commits you're expecting, make sure the repo is up to date.
+/// This function doesn't include the `since_commit` in the list since that commit was synced last time
+/// this tool was run.
+fn commits_to_be_applied(smithy_rs_repo: &Repository, since_commit: &Oid) -> Result<Vec<Oid>> {
+    let rev_range = format!("{}..HEAD", since_commit);
+    println!("Checking for smithy-rs commits in range {}", rev_range);
+
+    let mut rev_walk = smithy_rs_repo.revwalk().context(here!())?;
+    rev_walk.push_range(&rev_range).context(here!())?;
+
+    let mut commit_revs = Vec::new();
+    for rev in rev_walk {
```

Comment:
`rev_walk.map(...).collect()` maybe?

Reply:
I'm using a for loop so I can early return if there's an error when reading a rev

Reply:
An iterator over `Result<T, E>` can be `collect()`-ed directly into a `Result<Vec<T>, E>`, on which you can then use `?`, and it will return early upon encountering the first error.